hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/SnapshotVerifyProcedure.java

```diff
@@ -0,0 +1,182 @@
+import org.apache.hadoop.hbase.regionserver.SnapshotVerifyCallable;
+import org.apache.hadoop.hbase.snapshot.CorruptedSnapshotException;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.apache.hadoop.hbase.shaded.protobuf.ProtobufUtil;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProcedureProtos.SnapshotVerifyParameter;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProcedureProtos.SnapshotVerifyProcedureStateData;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.SnapshotProtos.SnapshotDescription;
+
+/**
+ *  A remote procedure which is used to send verify snapshot request to region server.
+ */
+@InterfaceAudience.Private
+public class SnapshotVerifyProcedure
+    extends ServerRemoteProcedure implements ServerProcedureInterface {
+  private static final Logger LOG = LoggerFactory.getLogger(SnapshotVerifyProcedure.class);
+
+  private SnapshotDescription snapshot;
+  private List<RegionInfo> regions;
+  private int expectedNumRegion;
+  private CorruptedSnapshotException remoteException;
+
+  public SnapshotVerifyProcedure() {}
+
+  public SnapshotVerifyProcedure(SnapshotDescription snapshot, List<RegionInfo> regions,
+      ServerName targetServer, int expectedNumRegion) {
+    this.targetServer = targetServer;
+    this.snapshot = snapshot;
+    this.regions = regions;
+    this.expectedNumRegion = expectedNumRegion;
+  }
+
+  @Override
+  protected void complete(MasterProcedureEnv env, Throwable error) {
+    if (error != null) {
+      Throwable realError = error.getCause();
+      if (realError instanceof CorruptedSnapshotException) {
+        synchronized (this) {
+          this.remoteException = (CorruptedSnapshotException) realError;
+        }
+        this.succ = true;
+      } else {
+        this.succ = false;
+      }
+    } else {
+      this.succ = true;
+    }
+  }
+
+  @Override
+  protected synchronized Procedure<MasterProcedureEnv>[] execute(MasterProcedureEnv env)
+      throws ProcedureYieldException, ProcedureSuspendedException, InterruptedException {
+    // Regardless of success or failure, ServerRemoteProcedure returns and leaves the parent
+    // procedure to find out and handle failures. In this case, SnapshotProcedure doesn't
+    // care which region server the task is assigned to, so we push down the choice of
+    // new target server to SnapshotVerifyProcedure.
+    Procedure<MasterProcedureEnv>[] res = super.execute(env);
+    if (res == null) {
+      if (succ) {
+        // remote task has finished, we already known snapshot if snapshot is corrupted
+        if (remoteException != null) {
+          setFailure("verify-snapshot", remoteException);
+        }
+        return null;
+      } else {
+        // can not send request to remote server, we will try another server
+        ServerName newTargetServer = env.getMasterServices().getServerManager().randomSelect();
```

Comment:
I think we'd better just fail here and let the upper layer to reschedule a procedure for verifying?

Reply:
That's a good idea. But I did not find a good way to let the parent procedure find that the child procedure failed to execute (As I mentioned in the doc, I think it's hard to communicate between parent procedure and child procedure), So I have to use some tricks to let the veiry procedure handle dispatch failure and keep retry until succeed. Do you have any ideas about how to  et the parent procedure find that the child procedure failed to execute ?